Improve accessibility of expander links, refs #13535

mInnes-archives opened the pull request:
PR adds aria-labels ~~and titles~~ to expander more/less links. ~~Titles add hover text to the "»" and "«" icons for general clarity~~. The aria-label redirects screen readers to ignore the link text, which currently reads as "right-pointing double arrow link" or some variation of this.

Testing utilised Orca and NVDA screen readers.

The only caveat with the added title is that it doesn't work with Symfony's i18n dictionary. However, browser-based translation picks up the slack.

jraddaoui reviewed (changes requested):
Thanks @mInnes-archives, nice addition!

A couple of details ...

**Translations:** we don't have a system to translate Javascript content in AtoM yet and those labels should be translated somehow. To work around this issue, we normally use data attributes in the related elements to pass the content translated to the JS code, see an example in the clipboard menu:

https://github.com/artefactual/atom/blob/qa/2.x/apps/qubit/modules/menu/templates/_clipboardMenu.php#L3-L7

In this case is hard to know where the expander will be added and therefore where to include those data attributes, as it may also happen in many pages. The following is not ideal in terms of performance but I'd suggest this solution for this particular case (and maybe others we're not aware) ...

[At the end of the footer](https://github.com/artefactual/atom/blob/qa/2.x/apps/qubit/templates/_footer.php#L24), add the following block:

```
<div id="js-i18n">
  <div id="read-more-less-links"
       data-read-more-text="<?php echo __('Read more'); ?>" 
       data-read-less-text="<?php echo __('Read less'); ?>">
  </div>
</div>
```

And from the JS:

```javascript
var $i18n = $('#js-i18n #read-more-less-links');
$element.find('.read-more a').attr('aria-label', $i18n.data('read-more-text'));
$element.find('.read-less a').attr('aria-label', $i18n.data('read-more-less'));
```

**Expanded status:** it would be great if we could add [`aria-expanded` attributes](https://www.w3.org/TR/wai-aria-1.1/#aria-expanded) to represent the status. I don't think it's a huge deal in this case and it doesn't have to happen in this PR, just a comment.

jraddaoui commented:
Some notes about merging this changes @mInnes-archives ...

Please, use the same ticket for the reference after rebasing and squashing the changes (https://projects.artefactual.com/issues/13535). If you submit all the pull requests ready to merge at the same time we'll need to rebase the others after merging one of them to be able to merge all with `--ff-only`. With our current infrastructure we have two options to solve this issue:

- If you care about being the author and committer (like this commit: 2c8f751), I'll need to ask you to rebase the other PRs and re-submit them after merging one; in which case we better do the rebase and merge one by one.
- If you are okay with being the author but not the committer (like: 6629c1f), feel free to submit all PRs at the same time and I'll take care of the rebase and merge.

mInnes-archives commented:
@jraddaoui, we'll do it one-by-one. This gives me time to implement the requested changes. I'll keep an eye on each as they get merged and set up the next if it's ready.

mInnes-archives commented:
@jraddaoui, I've added and tested your suggestion for the i18n data and everything works as it should. I've rebased and prepped for merge.

>Expanded status: it would be great if we could add aria-expanded attributes to represent the status.

In this case, I don't think this would be necessary/appropriate. The more and less links are technically two different elements, and their labeling is descriptive enough to convey the expanded status. Additionally, the SR output for the read-less link with `aria-expanded='true'` is a bit confusing. It reads things like 'read less link expanded' (Orca) or 'read less expanded link' (NVDA).

jraddaoui commented:
> @jraddaoui, I've added and tested your suggestion for the i18n data and everything works as it should. I've rebased and prepped for merge.
> 
> > Expanded status: it would be great if we could add aria-expanded attributes to represent the status.
> 
> In this case, I don't think this would be necessary/appropriate. The more and less links are technically two different elements, and their labeling is descriptive enough to convey the expanded status. Additionally, the SR output for the read-less link with `aria-expanded='true'` is a bit confusing. It reads things like 'read less link expanded' (Orca) or 'read less expanded link' (NVDA).

Thanks for looking into this Matt, sounds great!

The e2e tests are failing in the CI process, I took a look locally and found the following error in the console, which is preventing the addition of the `aria-label` attributes ...

```
TypeError: $ is not a function at HTMLDivElement.<anonymous> (qubit.js:34)
```

That JS is not wrapped in the usual ...

```js
(function($)
{
  'use strict';

  $(function()
  {
    
  });

})(jQuery);
```

Also, this requires another rebase, after d5c1097 (and another commit that I'll merge in a minute).

mInnes-archives committed:
Improve expander link accessibility, refs #13535

Squashed commit of the following:

commit 9c9341d6fe406ae8f882ddcde6fc9e9bba0ae8fc
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Wed Jul 7 13:02:47 2021 -0400

    Restructure function

commit 547d4582fcb45b77f921b91b2dc45b098d7eae6b
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Wed Jul 7 12:25:32 2021 -0400

    Wrap aria-label code in anon function

commit 5b16c8fd78460a1ea22d2bf19c77936a9c7adcc9
Author: Matt Innes <matthew.f.innes@gmail.com>
Date:   Sat Jul 3 22:29:35 2021 -0400

    Improve expander link accessibility, refs #13535

    Squashed commit of the following:

    commit 1160e0f787bb738d995ee961ad618a3597ad61f7
    Author: Matt Innes <matthew.f.innes@gmail.com>
    Date:   Sat Jul 3 21:57:42 2021 -0400

        i18n for read more/less link

    commit 482f7a4e154f0ecdaf75a62790595894b2cd4cb4
    Author: Matt Innes <matthew.f.innes@gmail.com>
    Date:   Sat Jun 12 11:55:28 2021 -0400

        Remove titles

    commit e7339124247f3373da8e259be49843eb90cc818e
    Author: Matt Innes <matthew.f.innes@gmail.com>
    Date:   Wed Jun 9 21:27:40 2021 -0400

        Correct variable names

    commit 5d1ff69c0a1e261a3fba97b254ce5f376a9546cb
    Author: Matt Innes <matthew.f.innes@gmail.com>
    Date:   Tue Jun 8 23:52:13 2021 -0400

        Improve accessibility of expander links

        - Add titles and arias labels to read more/less links

mInnes-archives commented:
Thanks @jraddaoui. It looks like that did the trick. I'll prep #1315 once this is merged.

jraddaoui commented:
Thanks @mInnes-archives!
